docs: Workflow Streams Python guide and encyclopedia concept page (#4513)
temporalio/documentation: docs/develop/python/workflows/workflow-streams.mdx

    The fix is to make the handler `async` and `await asyncio.sleep(0)` before reading state. That `yield` lets pending publishes apply with no Temporal timer or history events. Don't substitute `workflow.sleep(0)`, which schedules a timer. Most handlers are already safe like, anything that `await`s before reading state, the module's own poll handler, and Workflow-internal publishes.
- **Response size cap.** Poll responses are capped at roughly 1 MB. Subscribers repoll immediately when the server caps a response. No special handling is required, but high-throughput producers should expect a steady stream of small batches rather than occasional large ones.

## See also

- [Workflow Streams samples (samples-python)](https://github.com/temporalio/samples-python/tree/main/workflow_streams) — four runnable scenarios covering basic publish/subscribe, reconnecting subscribers, external publishers, and bounded logs via `truncate()`.

> - [`temporalio.contrib.workflow_streams` API reference](https://python.temporal.io/temporalio.contrib.workflow_streams.html).

Reviewer: This link goes to a 404.

Author: This is live now—following the release.

:::tip SUPPORT, STABILITY, and DEPENDENCY INFO

The `temporalio.contrib.workflow_streams` module is currently in
[Public Preview](/evaluate/development-production-features/release-stages#public-preview). Refer to the
[Temporal product release stages guide](/evaluate/development-production-features/release-stages) for more information.

Cross-language client support is on the roadmap. Only the Python client is available today.

The API may change before general availability.

:::

**Looking for...**

- Runnable end-to-end samples (basic publish/subscribe, reconnecting subscriber, external publisher, bounded log): [Workflow Streams samples](https://github.com/temporalio/samples-python/tree/main/workflow_streams).
- A complete LLM streaming example on this page (Activity publishes deltas, terminal consumer that resets on retry): [Stream LLM output](#stream-llm-output).
- Delivery guarantees, ordering, and retry semantics: [Delivery semantics](#delivery-semantics).
- History-size cost and tuning: [Architecture](#architecture).
- Long-running streams that need Continue-As-New: [Continue-As-New](#continue-as-new).

> ## Concepts

Reviewer 2: Wanted to flag this concern one more time. I am assuming when the other languages come in we'll want to move this content out even if you keep it here today. 

The way it is now, we'll end up with Python specific links--all external inks used in marketing materials--that we will have a hard time redirecting because this page will still be live. If you are good with that, go head and merge. Otherwise if it's not too much work, I'd still do the work now to write the conceptual docs in a language agnostic way to avoid headaches later.

Reviewer 3: Will follow up with an encyclopedia page in a separate PR!

---
id: workflow-streams
title: Workflow Streams - Python SDK
sidebar_label: Workflow Streams
toc_max_heading_level: 4
keywords:
  - workflow streams
  - streaming
  - signals
  - updates
  - queries
  - python sdk
tags:
  - Workflows
  - Messages
  - Streaming
  - Python SDK
  - Temporal SDKs
description: Stream events from a Workflow to subscribers using the Temporal Python SDK Workflow Streams contrib module.
---

**Workflow Streams** is a Temporal Python SDK `contrib` library that gives a Workflow a durable, offset-addressed event channel built on Temporal's basic message primitives: Signals, Updates, and Queries.

> It batches publishes to amortize per-Signal cost, deduplicates batches for exactly-once delivery to the log, supports topic filtering, and carries stream state across Continue-As-New for long-running runs.

Reviewer 3: Awkward wording
```suggestion
It batch-publishes events to amortize per-Signal cost, deduplicates batches for exactly-once delivery to the log, supports topic filtering, and carries state across Continue-As-New for long-running streams.
```

> Use Workflow Streams when you want outside observers to follow the progress of a Workflow and its Activities: updating a UI as an AI agent works, surfacing status from a payment or order pipeline, reporting intermediate results from a data job. It is not suited to ultra-low-latency cases like real-time voice, and it targets modest fan-out: tens of publishers and subscribers per Workflow, not thousands.

Reviewer 3: ```suggestion
Use Workflow Streams when you want outside observers to follow the progress of a Workflow and its Activities: updating a UI as an AI agent works, surfacing status from a payment or order pipeline, or reporting intermediate results from a data job. It is not suited to ultra-low-latency cases like real-time voice, and it targets modest fan-out: tens of publishers and subscribers per Workflow, not thousands.
```

:::tip SUPPORT, STABILITY, and DEPENDENCY INFO

The `temporalio.contrib.workflow_streams` module is currently in
[Public Preview](/evaluate/development-production-features/release-stages#public-preview). Refer to the
[Temporal product release stages guide](/evaluate/development-production-features/release-stages) for more information.

Cross-language client support is on the roadmap. Only the Python client is available today.

The API may change before general availability.

:::

**Looking for...**

- Runnable end-to-end samples (basic publish/subscribe, reconnecting subscriber, external publisher, bounded log): [Workflow Streams samples](https://github.com/temporalio/samples-python/tree/main/workflow_streams).
- A complete LLM streaming example on this page (Activity publishes deltas, terminal consumer that resets on retry): [Stream LLM output](#stream-llm-output).
- Delivery guarantees, ordering, and retry semantics: [Delivery semantics](#delivery-semantics).
- History-size cost and tuning: [Architecture](#architecture).
- Long-running streams that need Continue-As-New: [Continue-As-New](#continue-as-new).

> ## Concepts

Reviewer 2: We isolate conceptual content to encyclopedia pages. SDK guides need to stay task focused.

The Workflow keeps an in-memory append-only log of `(topic, data)` entries, each with a monotonically increasing offset. Subscribers maintain their own cursor and receive the next range past that cursor on every poll. See [Architecture](#architecture) for batching, dedup, history-size implications, and tuning.

When events originate in an Activity, for example an Activity streaming output from an LLM, have the Activity publish to the stream directly. The Workflow acts as a conduit: it hosts the stream, processes only the Activity's return value, and emits its own lifecycle events, but does not read the stream itself. This keeps the Workflow's own state independent of partial output from Activity attempts that ended up retrying, since the Workflow only sees the successful attempt's return. See [Delivery semantics](#delivery-semantics) for the precise guarantees.

## Where the stream lives

A `WorkflowStream` is hosted inside a Workflow, so the first design choice is whether one Workflow handles both the work and the stream, or whether a separate Workflow exists only to host the stream. The choice is mostly about lifecycle.

> **Host the stream on the Workflow that does the work** when the events come from what that Workflow is already orchestrating: an agent run, an order pipeline, a chat session. The stream's lifecycle aligns with the run, starting when the run starts and ending when it returns. The Workflow Id you use to start the work is the same one subscribers attach to. This is the common shape for AI agents and most progress-streaming cases, where streaming is just one more thing the Workflow does as part of its job.

Reviewer 3: ```suggestion
**Host the stream on the Workflow that does the work** when the events come from what that Workflow is already orchestrating: an agent run, an order pipeline, a chat session. The stream's lifecycle aligns with the run, starting when the run starts and ending when it returns. The Workflow ID you use to start the work is the same one subscribers attach to. This is the common shape for AI agents and most progress-streaming cases, where streaming is just one more thing the Workflow does as part of its job.
```

**Use a dedicated Workflow for the stream alone** when the stream should outlive any single producer, accept fan-in from multiple unrelated sources, or be subscribable before any work has started. Producers publish from outside the stream Workflow (Activities of other Workflows, or external `WorkflowStreamClient` instances). The trade-off is explicit lifecycle management: a dedicated stream Workflow does not terminate on its own, so you need a signal-driven shutdown or a Continue-As-New strategy.

> Whichever shape you pick, the Workflow Id is the address subscribers use to attach. Multiple subscribers can attach to the same Id concurrently, which is the normal case for a UI with multiple browser tabs. Use distinct Workflow Ids for unrelated streams rather than packing them into one Workflow.

Reviewer 3: ```suggestion
Whichever shape you pick, the Workflow ID is the address subscribers use to attach. Multiple subscribers can attach to the same ID concurrently, which is the normal case for a UI with multiple browser tabs. Use distinct Workflow IDs for unrelated streams rather than packing them into one Workflow.
```

        self.status = self.stream.topic("status", type=StatusEvent)

    @workflow.run
    async def run(self, input: OrderInput) -> None:
        self.status.publish(StatusEvent(state="validating", detail="checking inventory"))
        await validate_order(input.order_id)

        self.status.publish(StatusEvent(state="charging", progress=33, detail="authorizing payment"))
        await charge_payment(input.order_id)

        self.status.publish(StatusEvent(state="shipping", progress=66, detail="dispatching to warehouse"))
        await dispatch_order(input.order_id)

        self.status.publish(StatusEvent(state="completed", progress=100))
```

`publish()` runs the payload converter to encode each value. The codec chain (encryption, compression, and so on) runs once on the Signal or Update envelope that carries the batch, never per item, so encryption and compression are applied exactly once each direction.

The `type=` argument is optional and defaults to `Any`. Pass it when you want the binding recorded so re-binding the same name to an unequal type raises, or so subscribers can pick up the type from the same handle.

## Publish from a client

> Any process that has a Temporal `Client` and the target Workflow Id can publish to that Workflow's stream by constructing a `WorkflowStreamClient`. This is the general pattern and covers HTTP backends, starters, one-off scripts, other Workflows' Activities, and standalone Activities. Construct one with `WorkflowStreamClient.create(client, workflow_id)`, then use it the same way you would the Workflow-side handle: bind a topic, publish through it, and let the async context manager flush on exit.

Reviewer 3: ```suggestion
Any process that has a Temporal `Client` and the target Workflow ID can publish to that Workflow's stream by constructing a `WorkflowStreamClient`. This is the general pattern and covers HTTP backends, starters, one-off scripts, other Workflows' Activities, and standalone Activities. Construct one with `WorkflowStreamClient.create(client, workflow_id)`, then use it the same way you would the Workflow-side handle: bind a topic, publish through it, and let the async context manager flush on exit.
```

```python
from datetime import timedelta

from temporalio.client import Client
from temporalio.contrib.workflow_streams import WorkflowStreamClient

async def publish_status(workflow_id: str) -> None:
    temporal_client = await Client.connect("localhost:7233")
    stream_client = WorkflowStreamClient.create(
        temporal_client,
        workflow_id=workflow_id,
        batch_interval=timedelta(milliseconds=200),
    )
    async with stream_client:
        status = stream_client.topic("status", type=StatusEvent)
        status.publish(StatusEvent(state="started"))
        ...
        # Buffer is flushed on context manager exit.
```

> Inside an Activity scheduled by a Workflow, `WorkflowStreamClient.from_within_activity()` is a convenience that infers the Temporal `Client` and the parent Workflow Id from the Activity context, so you don't have to thread them through the Activity's input:

Reviewer 3: ```suggestion
Inside an Activity scheduled by a Workflow, `WorkflowStreamClient.from_within_activity()` is a convenience that infers the Temporal `Client` and the parent Workflow ID from the Activity context, so you don't have to thread them through the Activity's input:
```

    async with stream_client:
        deltas = stream_client.topic("delta", type=TextDelta)
        complete = stream_client.topic("complete", type=TextComplete)
        retry = stream_client.topic("retry", type=dict)

        # Tell consumers an earlier attempt's deltas are stale.
        if activity.info().attempt > 1:
            retry.publish({"attempt": activity.info().attempt}, force_flush=True)

        full: list[str] = []
        first = True
        oai_stream = await openai_client.chat.completions.create(
            model="gpt-4o-mini",
            messages=[{"role": "user", "content": prompt}],
            stream=True,
        )
        async for chunk in oai_stream:
            if not chunk.choices:
                continue
            text = chunk.choices[0].delta.content
            if not text:
                continue
            # force_flush only on the first delta so the user sees something
            # immediately; subsequent deltas batch at the 200 ms interval.
            deltas.publish(TextDelta(text=text), force_flush=first)
            first = False
            full.append(text)
        complete.publish(TextComplete(full_text="".join(full)))

>     return "".join(full)

Reviewer 3: Why is it useful to publish the full text if the Activity also returns it?

Author: Let's delete this - activity can return.
